python/lsst/obs/metadata/metadata.py

```diff
@@ -0,0 +1,179 @@
+            Header values representing a FITS header.
+
+        Returns
+        -------
+        translation : `str` or `numbers.Number`
+            Always returns {constant}.
+        """
+        return translator
+
+    @staticmethod
+    def _makeUnitMapping(standardKey, fitsKey):
+        def translator(header):
+            return header[fitsKey]
+        translator.__doc__ = f"""Map '{fitsKey}' FITS keyword to '{standardKey}' property
+
+        Parameters
+        ----------
+        header : `dict`-like
+            Header values representing a FITS header.
+
+        Returns
+        -------
+        translation : `str` or `numbers.Number`
+            Translated header value derived directly from the {fitsKey}
+            header value.
+        """
+        return translator
+
+    def __init__(self, name, bases, dct):
+        super().__init__(name, bases, dct)
+
+        if hasattr(self, "name") and self.name is not None:
+            MetadataTranslator.translators[self.name] = self
+
+        for standardKey, fitsKey in self._unitMap.items():
+            translator = self._makeUnitMapping(standardKey, fitsKey)
+            setattr(self, f"to{standardKey.capitalize()}", staticmethod(translator))
+
+        for standardKey, constant in self._constMap.items():
+            translator = self._makeConstMapping(standardKey, constant)
+            setattr(self, f"to{standardKey.capitalize()}", staticmethod(translator))
+
+
+class MetadataTranslator(metaclass=MetadataMeta):
+    """Per-instrument metadata translation support"""
+
+    _unitMap = {}
+    """Dict of one-to-one mappings for header translation from standard
+    property to corresponding FITS keyword."""
+
+    _constMap = {}
+    """Dict defining a constant for specified standard properties."""
+
+    translators = dict()
+    """All registered metadata translation classes."""
+
+    supportedInstrument = None
+    """Name of instrument understood by this translation class."""
+
+    @classmethod
+    @abstractmethod
+    def canTranslate(cls, header):
+        """Indicate whether this translation class can translate the
+        supplied header.
+
+        Parameters
+        ----------
+        header : `dict`-like
+           Header to convert to standardized form.
+
+        Returns
+        -------
+        can : `bool`
+            `True` if the header is recognized by this class. `False`
+            otherwise.
+        """
+        raise NotImplementedError()
+
+
+class VisitInfo:
+    """Standardized representation of an instrument FITS header.
+
+    Parameters
+    ----------
+    header : `dict`-like
+        Representation of an instrument FITS header accessible as a `dict`.
+    translator : `MetadataTranslator`-class, `optional`
+        If not `None`, the class to use to translate the supplied headers
+        into standard form. Otherwise each registered translator class will
+        be asked in turn if it knows how to translate the supplied header.
```

Comment:
I'm not opposed to this functionality (being able to determine what translators to use for a given header), but I think we'd be able to explicitly declare what translator(s) to use for any header.  I'm not suggesting that we should, but if this does become problematic for some reason, I think we can live without it.

Reply:
I don't like the sound of per-header translator as you've written above but maybe I don't understand what you mean.  This approach worked well for me at $OTHER_TELESCOPE and you can override the translator class in the constructor.  In my perl implementation we had about 20 instruments and that included cases where the instrument header changed fundamentally on a certain date and this the `canTranslate` methods would take the date into account and give you the right collection of translation methods.

ORAC-DR had a fully self-describing pipeline configuration system so the data file was discovered and a priori you didn't know what it was so the first step was to ask the file to translate itself and return the standard information. At that point the instrument-speicifc configurations (pipeline recipes etc) could be loaded.  File ingest in the butler could be done the same way. You don't want to be forced to declare that the files being read are LSSTCam or HSC files; reading the metadata is enough for you to know what to do next.

Reply:
I was actually using "per-header translator" to describe how I perceived *this* system, just because you pass (or find) *one* translator when constructing a `VisitInfo` from a header.

Anyhow, I agree there's nothing obviously problematic here.  I just wanted to point out that our use cases probably could always provide a not-`None` argument for the `translator` argument.

Reply:
Right. "header" is overloaded and can mean the whole header or a header card. Sorry. My vision is not that you try to cobble together a bunch of translator classes and throw them at the `VisitInfo` constructor. You cobble together a single translator class out of the component translators and then pass that single coherent entity to the constructor.  Declaring that a header from a file is a bit of decam and a bit of HSC should be done at the translator class level.  Otherwise you risk getting two `to_property` methods and not being entirely sure which one should be used.  At the class definition level you can remove all the ambiguity and run unit tests to ensure it all works.

Headers from instruments are meant to be predictable. They are also meant to be self-describing (at least to the extent of unambiguously being able to work out what instrument they are).